numba/typing/builtins.py

```diff
@@ -726,6 +726,11 @@ def resolve_stop(self, ty):
     def resolve_step(self, ty):
         return types.intp
 
+    @bound_function("slice.indices")
+    def resolve_indices(self, ty, args, kws):
+        assert not kws
+        return signature(types.UniTuple(types.intp, 3), types.intp)
```

Comment:
This needs a guard to check that `args[0]` is a `types.Integer` and only 1 arg is present e.g.:
```python
        val = args[0]
        if len(args==1) and isinstance(val, types.Integer):
            return signature(types.UniTuple(types.intp, 3), types.intp)
```
without this there's an implicit guarantee that this function will consume an `intp` and cast anything to it and return a 3-tuple. i.e. this "works", when it really should not.

```python
from numba import njit

a  = slice(1, 9, 2)

@njit
def foo(x):
    return x.indices(12.34)

print(foo(a))
```
as does this:
```python
from numba import njit

a  = slice(1, 9, 2)

@njit
def foo(x):
    return x.indices(10, 20, 30, 40, 50, 60)

print(foo(a))
```
both of the above perhaps ought to become part of the unit tests?

Reply:
Ah, now I see what's up with the convention of `[ty] = args`.

Should these throw typing errors, or should I try to match the python errors?

Reply:
I'm pretty sure they won't need to explicitly throw anything, if a typing template doesn't return a signature it means the type combination being compiled isn't supported (and an automatically generated TypingError should appear). Perhaps check that is the case first and if the error message makes sense, if so, it's just a question of adding tests to make sure that that happens correctly. If the messages are weird it's fine to instead have an `else` branch raise a TypingError with a specific message that you feel is more suitable.

Reply:
Based on a few other `bound_function` definitions ([example](https://github.com/numba/numba/blob/8c6658375c1f8fe50e1a5ccd11d4e7bf5a8053de/numba/typing/arraydecl.py#L299)) I tried raising the error python would have thrown. This gets overwritten with a typing error, but at least the message is included. Here's an reduced example of what this looks like:

```python
>>> from numba import njit
>>> slice_indices = lambda s, *indargs: s.indices(*indargs)
>>> cslice_indices = njit(slice_indices)

>>> slice_indices(slice(None), 1.2)
TypeError                                 Traceback (most recent call last)
...
TypeError: 'float' object cannot be interpreted as an integer

>>> cslice_indices(slice(None), 1.2)
TypingError                               Traceback (most recent call last)
...
TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Internal error at <numba.typeinfer.CallConstraint object at 0x625301e50>.
'float64' object cannot be interpreted as an integer
...
```

Thoughts?

Reply:
Seems good, thanks.